### ISSUE TITLE
Learners attempt to match each template only once

### ISSUE THREAD

**Issue by spigo900** [isi-vista/adam#997]
Our template learners (excepting for the object learner in #987) only attempt to match each template once. This means we get at most one semantic node (and one description) per template.

We need to fix this so that we can handle things like the relation "on." In the scene "a ball on a table" it's implicit that the table is on the ground, so the learner might also match its `slot1_on_slot2` template with "a table on the ground" instead of "a ball on a table." Ideally we'd like the relation learner to be able to identify both of these relations, but currently it can only identify one or the other.

This might also be needed for experiments with attribute/relation noise.

## Solutions?
We don't have a general fix for this yet. @denizbeser and I fixed this for the object learner in #987, however the fix depends on the fact that the object learner replaces its matches. Other learners don't and shouldn't do this, because they would remove information later learners need. For example, the relation learner might identify the relation "a dog on the ground" and replace the dog, which prevents the action learner from later recognizing "a dog eats a cookie."

One idea is to simply match the templates more than once. The problem with this is that we will may end up with extra matches (and so semantic nodes) for the "same things" (attribute, relation, verb) where the differences are trivial. For example, we might get extra matches for a verb template where the only difference is how we matched up the predicates for the world's horizontal axes differently. These extra matches won't get thrown away when we take the set of `immutable_new_nodes` because attribute, relation, and action semantic nodes are `eq=False`. And because they won't get thrown away, they might slow down learners later in the learning order.

We could fix the problems with that idea by changing the semantic nodes to be `eq=True`. I can only see two minor problems with that:

1. Unless we add some information to these nodes, we won't be able to recognize multiple distinct instances of the same action in the same scene (if, for example, we had a situation where Mom sits down, then gets up, then sits down again).
    - However, we currently don't have any such situations and aren't likely to need them in the time we have left, so I think this is fine.
2. It might slow things down since we will be looking for every match, not just the first.
    - I think this is probably fine as well.

**Comment by spigo900** [comment]
@LichMaster98 If I am looking at the same problem as the one you [mentioned](https://usc-isi.slack.com/archives/GL17LRSTG/p1605553302246000), I think this is what's blocking the pursuit relations learner (#990) from recognizing "a ball on a table."

**Comment by LichMaster98** [comment]
@spigo900 This issue makes a lot of sense yes and would explain what I was seeing earlier. For the problems listed, case 1 is I believe out of scope in our current implementations. We aren't planning on having multiple 'stages' of actions so Mom can only sit down once in an observation. 

I think we'll just have to take the speed penalty by searching for every match and not just the first because we need the ability to continue to explore the search space to gather the examples rather than stopping with the first accurate observation of the scene.
